phtree/common/filter.h

```diff
@@ -150,6 +150,132 @@ class FilterAABB {
+        KeyExternal point = converter_.post(key);
+
+        // radius of paraboloid at axis offset of point
+        ScalarExternal diff_axis = point[axis_] - vertex_external_[axis_];
+        ScalarExternal radius = sqrt(diff_axis / scale_factor_);
+
+        // point projected onto axis of paraboloid
+        KeyExternal projected = vertex_external_;
+        projected[axis_] = point[axis_];
+
+        return distance_function_(projected, point) <= radius;
+    }
+
+    /*
+     * Calculate whether AABB encompassing all possible points in the node intersects with the
+     * paraboloid.
+     */
+    [[nodiscard]] bool IsNodeValid(const KeyInternal& prefix, int bits_to_ignore) const {
+        // we always want to traverse the root node (bits_to_ignore == 64)
+
+        if (bits_to_ignore >= (MAX_BIT_WIDTH<ScalarInternal> - 1)) {
+            return true;
+        }
+
+        ScalarInternal node_min_bits = MAX_MASK<ScalarInternal> << bits_to_ignore;
+        ScalarInternal node_max_bits = ~node_min_bits;
+
+        // coordinate on axis that is inside the paraboloid and furthest away from vertex
+        ScalarInternal coord_max_vertex_dist;
+        if (scale_factor_ < 0) {
+            // lower bound is furthest from the vertex
+            coord_max_vertex_dist = prefix[axis_] & node_min_bits;
+            if (coord_max_vertex_dist > vertex_internal_[axis_]) {
+                // point is on the side of the vertex where the paraboloid does not exist
+                return false;
+            }
+        } else {
+            // upper bound is furthest from the vertex
+            coord_max_vertex_dist = prefix[axis_] | node_max_bits;
+            if (coord_max_vertex_dist < vertex_internal_[axis_]) {
+                // point is on the side of the vertex where the paraboloid does not exist
+                return false;
+            }
+        }
+
+        KeyInternal closest_to_axis;
+        for (size_t i = 0; i < prefix.size(); ++i) {
+            if (i == axis_) {
+                closest_to_axis[i] = coord_max_vertex_dist;
+            } else {
+                // calculate lower and upper bound for dimension for given node
+                ScalarInternal lo = prefix[i] & node_min_bits;
+                ScalarInternal hi = prefix[i] | node_max_bits;
+
+                // choose value closest to vertex for dimension
+                closest_to_axis[i] = std::clamp(vertex_internal_[i], lo, hi);
+            }
+        }
+
+        // closest_to_axis projected onto axis of paraboloid
+        KeyExternal projected = vertex_internal_;
+        projected[axis_] = coord_max_vertex_dist;
+
+        KeyExternal closest_point = converter_.post(closest_to_axis);
+        KeyExternal projected_point = converter_.post(projected);
+
+        // radius of paraboloid at axis offset of closest_to_axis
+        ScalarExternal diff_axis = closest_point[axis_] - vertex_external_[axis_];
+        ScalarExternal radius = sqrt(diff_axis / scale_factor_);
```

Comment:
I am not sure it makes a difference, but it may:
The sqrt could be avoided by comparing the squared radius. Division could be avoided by storing the inverse `scale_factor_`

Reply:
Good point, is updated.